Objects/longobject.c

```diff
@@ -1758,6 +1759,20 @@ long_to_decimal_string_internal(PyObject *aa,
     size_a = Py_ABS(Py_SIZE(a));
     negative = Py_SIZE(a) < 0;
 
+    /* quick and dirty pre-check for overflowing the digit limit,
+       based on the inequality 10/3 >= log2(10) */
+    if (size_a >= 10 * _PY_LONG_MAX_STR_DIGITS_THRESHOLD
+                  / (3 * PyLong_SHIFT) + 2) {
+        PyInterpreterState *interp = _PyInterpreterState_GET();
+        int max_str_digits = interp->int_max_str_digits;
+        if ((max_str_digits > 0) && (size_a >= 10 * max_str_digits
```

Comment:
The multiplication in this line is potentially problematic: `max_str_digits` has type `int`, and `10 * max_str_digits` could potentially overflow. We may need to impose a maximum for `max_str_digits`.

Separately from that, `int` seems a little small for `max_str_digits` - `INT_MAX` could potentially be as small as `32767`, meaning that to avoid overflow in `10 * max_str_digits` we'd need `max_str_digits <= 3276`. Would it make sense to use either `Py_ssize_t` or a fixed-width type like `int32_t` or `int64_t` instead?

Reply:
I've updated to include an overflow check. This way, if `int_max_str_digits` is huge, the code behaves as though there's no limit. That may not be ideal, but it's better than having undefined behaviour.